postgres/__init__.py

```diff
@@ -838,6 +840,30 @@ def make(self, values):
             instance = ModelSubclass(record)
             return instance
 
+        def parse(self, s, curs, retry=False):
+            # Override to protect against race conditions:
+            #   https://github.com/gratipay/postgres.py/issues/26
+
+            if s is None:
+                return None
+
+            tokens = self.tokenize(s)
+            if len(tokens) != len(self.atttypes):
+                # The number of columns has changed, re-fetch the type info
+                self.__dict__.update(self._from_db(self.name, curs).__dict__)
```

Comment:
Here's [`CompositeCaster` at `HEAD` of `master`](https://github.com/psycopg/psycopg2/blob/5efe7131ff86ba5adf08ead91393f4910e863f1e/lib/extras.py#L805-L938).